vncviewer/EmulateMB.cxx

```diff
@@ -194,23 +205,198 @@ static const signed char stateTab[11][5][3] = {
   },
 };
 
-EmulateMB::EmulateMB()
-  : state(0), emulatedButtonMask(0), timer(this)
+/*
+ * 2) Pressing a modifier + left button to emulate middle button:
+ *
+ * At each state, we need handlers for the following events
+ *   0: Modifier pressed
+ *   1: Modifier released
+ *   2: Left button pressed / released
+ *   3: Other button pressed / released
+ *   4: A key is pressed/released (not the modifier)
+ *
+ * a handler consists of three values:
+ *   0: action1 - Modifler to send
+ *   1: action2 - Button emulation mode
+ *   2: Next emulation state
+ *
+ * action1:
+ * action = 2: Send modifier down followed by modifier up
+ * action = 1: Send modifier down
+ * action = 0: nothing
+ * action = -1: Send modifier up
+ * action = 10: print error, nothing
+ *
+ * action2:
+ * action = 0: Dont Emulate
+ * action < 0: Emulate
+ *
+ * Note that a key pressed event (where the key is not the emulation modifier) always sends the key.
+ */
+
+EmulateMB::Action EmulateMB::mbStateTab[8][5] = {
+/* 0 ground */
+  {
+    {  0,  0,  1 },   /* mod down -> mod down */
```

Comment:
Is this how other such systems behave? I would suspect that pressing Alt here and then releasing it on left mouse down would be more robust.